File: archive/cdk-py/Pr2081/lib/database_stack.py
"""database_stack.py
RDS database configuration with proper security and backup settings.
"""

import aws_cdk as cdk
from aws_cdk import aws_rds as rds, aws_ec2 as ec2
from constructs import Construct


class DatabaseStack(cdk.NestedStack):
    """Creates RDS database with production-ready configuration."""

    def __init__(
        self,
        scope: Construct,
        construct_id: str,
        vpc: ec2.Vpc,
        database_security_group: ec2.SecurityGroup,
        environment_suffix: str,
        **kwargs
    ):
        super().__init__(scope, construct_id, **kwargs)

        # Database subnet group
        db_subnet_group = rds.SubnetGroup(
            self, f"prod-db-subnet-group-{environment_suffix}",
            description="Subnet group for RDS database",
            vpc=vpc,
            vpc_subnets=ec2.SubnetSelection(
                subnet_type=ec2.SubnetType.PRIVATE_ISOLATED
            ),
            subnet_group_name=f"prod-db-subnet-group-{environment_suffix}"
        )

        # Database credentials
        db_credentials = rds.Credentials.from_generated_secret(
            username="admin",
            secret_name=f"prod-db-credentials-{environment_suffix}"
        )

        # RDS Instance
        self.database = rds.DatabaseInstance(
            self, f"prod-database-{environment_suffix}",
            engine=rds.DatabaseInstanceEngine.mysql(
                version=rds.MysqlEngineVersion.VER_8_0_42
            ),
            instance_type=ec2.InstanceType.of(
                ec2.InstanceClass.BURSTABLE3,
                ec2.InstanceSize.MICRO
            ),
            credentials=db_credentials,
            vpc=vpc,
            subnet_group=db_subnet_group,
            security_groups=[database_security_group],
            database_name=f"prod_app_db_{environment_suffix}",
            allocated_storage=20,
            max_allocated_storage=100,
            storage_encrypted=True,
            multi_az=False,  # Single AZ for cost optimization with t3.micro
            backup_retention=cdk.Duration.days(7),
            delete_automated_backups=False,
            deletion_protection=False,  # Must be False for cleanup
            enable_performance_insights=False,  # Not available for t3.micro
            auto_minor_version_upgrade=True,
            parameter_group=rds.ParameterGroup.from_parameter_group_name(
                self, f"prod-db-params-{environment_suffix}",
                "default.mysql8.0"
            )
        )

        # Export database credentials secret ARN
        cdk.CfnOutput(
            self, "DatabaseCredentialsSecretArn",
            value=self.database.secret.secret_arn,
            description="ARN of the database credentials secret"
        )

File: archive/cdk-ts/Pr3242/lib/database-stack.ts
import * as cdk from 'aws-cdk-lib';
import * as ec2 from 'aws-cdk-lib/aws-ec2';
import * as kms from 'aws-cdk-lib/aws-kms';
import * as rds from 'aws-cdk-lib/aws-rds';
import * as s3 from 'aws-cdk-lib/aws-s3';
import * as secretsmanager from 'aws-cdk-lib/aws-secretsmanager';
import { Construct } from 'constructs';

interface DatabaseStackProps {
  vpc: ec2.Vpc;
  securityGroup: ec2.SecurityGroup;
  backupBucket: s3.Bucket;
  environmentSuffix: string;
}

export class DatabaseStack extends Construct {
  public readonly database: rds.DatabaseInstance;

  constructor(scope: Construct, id: string, props: DatabaseStackProps) {
    super(scope, id);

    // Get the current region for region-specific configuration
    const region = cdk.Stack.of(this).region;

    // Use simplest PostgreSQL configuration with default settings
    const databaseEngine = rds.DatabaseInstanceEngine.POSTGRES;

    // PostgreSQL 15.4 optimized for us-east-2 region

    // Output the selected database engine for debugging
    new cdk.CfnOutput(this, 'DatabaseEngineUsed', {
      value: `Database engine selected for region ${region}: PostgreSQL (Default Version)`,
      description:
        'Database engine automatically selected based on region compatibility',
    });

    // Create KMS key for database encryption
    const encryptionKey = new kms.Key(this, 'DatabaseEncryptionKey', {
      description: 'KMS key for RDS database encryption',
      enableKeyRotation: true,
      alias: `retail-db-key-${props.environmentSuffix}`,
      removalPolicy: cdk.RemovalPolicy.DESTROY, // Must be destroyable for testing
    });

    // Create database credentials secret
    const databaseCredentials = new secretsmanager.Secret(
      this,
      'DatabaseCredentials',
      {
        description: 'RDS PostgreSQL database credentials',
        generateSecretString: {
          secretStringTemplate: JSON.stringify({ username: 'dbadmin' }),
          generateStringKey: 'password',
          excludeCharacters: ' %+~`#$&*()|[]{}:;<>?!\'/@"\\',
          passwordLength: 32,
        },
        removalPolicy: cdk.RemovalPolicy.DESTROY, // Must be destroyable for testing
      }
    );

    // Create subnet group
    const subnetGroup = new rds.SubnetGroup(this, 'DatabaseSubnetGroup', {
      description: 'Subnet group for RDS PostgreSQL',
      vpc: props.vpc,
      vpcSubnets: {
        subnetType: ec2.SubnetType.PRIVATE_ISOLATED,
      },
      removalPolicy: cdk.RemovalPolicy.DESTROY, // Must be destroyable for testing
    });

    // Create RDS PostgreSQL instance
    this.database = new rds.DatabaseInstance(this, 'RetailDatabase', {
      instanceIdentifier: `retail-db-${props.environmentSuffix}`,
      engine: databaseEngine,
      instanceType: ec2.InstanceType.of(
        ec2.InstanceClass.T3,
        ec2.InstanceSize.MICRO
      ),
      vpc: props.vpc,
      subnetGroup: subnetGroup,
      securityGroups: [props.securityGroup],
      credentials: rds.Credentials.fromSecret(databaseCredentials),
      allocatedStorage: 20,
      maxAllocatedStorage: 100,
      storageType: rds.StorageType.GP3,
      storageEncrypted: true,
      storageEncryptionKey: encryptionKey,
      multiAz: false,
      autoMinorVersionUpgrade: true,
      backupRetention: cdk.Duration.days(7),
      preferredBackupWindow: '03:00-04:00',
      preferredMaintenanceWindow: 'sun:04:00-sun:05:00',
      deletionProtection: false, // Disabled for testing - must be destroyable
      removalPolicy: cdk.RemovalPolicy.DESTROY, // Must be destroyable for testing
      enablePerformanceInsights: true,
      performanceInsightRetention: rds.PerformanceInsightRetention.DEFAULT,
      performanceInsightEncryptionKey: encryptionKey,
      monitoringInterval: cdk.Duration.seconds(60),
      databaseName: 'retaildb',
      // No custom parameter group - use AWS defaults for maximum compatibility
    });

    // Note: RDS automatic backups are handled by AWS internally
    // The S3 bucket is for additional manual backups if needed

    // Output database endpoint
    new cdk.CfnOutput(this, 'DatabaseEndpoint', {
      value: this.database.dbInstanceEndpointAddress,
      description: 'RDS database endpoint',
    });

    new cdk.CfnOutput(this, 'DatabasePort', {
      value: this.database.dbInstanceEndpointPort,
      description: 'RDS database port',
    });

    // Tag resources
    cdk.Tags.of(this).add('Environment', props.environmentSuffix);
    cdk.Tags.of(this).add('Application', 'RetailDatabase');
    cdk.Tags.of(this).add('ManagedBy', 'CDK');
  }
}
